### PR TITLE
More Refactors

### PR DESCRIPTION
**Pull request by ripytide**
I don't think I've taken the opportunity to thank you for this amazing utility so far, so thank you!

In this PR:
- Refactor the confusing/bodgy consts backend macro to use a single
  `BackendInfo` struct
- Combine `supports_as_dependency()` and the `switches_dependency` into a single `Option` type as that is a better representation as it is an optional backend feature.
- Add more types like `Packages`, `Sections` for uniformity.
- Added a crate-private prelude and switches all imports to use a
  prelude glob import (excluding functions because it's nice to only be
  importing types)
- Refactored Backend section name in logging by implementing Display for
  `AnyBackend` that just prints the `section`

### CODE REVIEW

File: crates/pacdef/src/main.rs [hunk 1]
```diff
@@ -10,8 +10,7 @@
     clippy::unwrap_used,
     clippy::use_debug,
     clippy::use_self,
-    clippy::wildcard_dependencies,
-    missing_docs
```
Comment (steven-omaha) [inline crates/pacdef/src/main.rs:14]: Keep the lints. If you'd rather not write the docstrings yourself I will take care of them.

File: crates/pacdef/src/lib.rs [hunk 2]
```diff
@@ -16,12 +16,10 @@
     clippy::unwrap_used,
     clippy::use_debug,
     clippy::use_self,
-    clippy::wildcard_dependencies,
-    missing_docs
```
Comment (steven-omaha) [inline crates/pacdef/src/lib.rs:20]: See main.rs

File: crates/pacdef/src/backend/backend_trait.rs [hunk 3]
```diff
@@ -1,82 +1,72 @@
...
-    /// Return the default binary as defined in the constant of the module.
-    /// See [`Backend::get_binary()`].
-    fn get_binary_default(&self) -> Text;
-
-    /// Get the name of the section in the group files.
-    fn get_section(&self) -> Text;
-
-    /// Get CLI switches for the package manager to show information for
+// A bundle of small of bits of info associated with a backend.
+pub struct BackendInfo {
```
Comment (steven-omaha) [inline crates/pacdef/src/backend/backend_trait.rs:15]: Make this a docstring

File: crates/pacdef/src/backend/mod.rs [hunk 4]
```diff
@@ -1,23 +1,53 @@
...
 
-use self::actual::{
-    fedora::Fedora, flatpak::Flatpak, python::Python, rust::Rust, rustup::Rustup, void::Void,
-};
+use crate::prelude::*;
+use anyhow::{Context, Result};
 
-#[derive(Debug)]
+// A backend with its associated managed packages
+pub struct ManagedBackend {
```
Comment (steven-omaha) [inline crates/pacdef/src/backend/mod.rs:12]: docstring

File: crates/pacdef/src/core.rs [hunk 5]
```diff
@@ -450,29 +423,40 @@ fn overwrite_values_from_config(backend: &mut AnyBackend, config: &Config) {
...
+            Ok(diff) => todo_unmanaged.push((any_backend.clone(), diff)),
+            Err(error) => show_backend_query_error(&error, any_backend),
+        };
 
-        match backend.get_unmanaged_packages_sorted() {
-            Ok(unmanaged) => result.push((backend, unmanaged)),
-            Err(error) => show_backend_query_error(&error, &backend),
+        match managed_backend.get_unmanaged_packages_sorted() {
+            Ok(unmanaged) => todo_unmanaged.push((any_backend.clone(), unmanaged)),
+            Err(error) => show_backend_query_error(&error, any_backend),
```
Comment (steven-omaha) [inline crates/pacdef/src/core.rs:455]: both code blocks are identical. I doubt that was the intention.

**Review by steven-omaha — CHANGES_REQUESTED** [review]
This is -- once again -- very good work. Most changes are small. 

If the docstring thingy fails the CI check, don't worry, I'll merge it regardless and add the docstrings soon.

There's two bugs:
- core.rs:448..456 are duplicated, one of the blocks needs to go
- there's a bug that I couldn't track down quickly, here's how to reproduce:

1. Have any Arch package with a repo prefix, like `core/linux`.
2. Run `cargo run -F arch -- p u`: it will show `linux` as unmanged
3. Run `cargo run -F arch -- p sy`: it will want to install `core/linux`.

That's not present in the previous commit, please check that out.

**Comment by steven-omaha** [comment]
> I don't think I've taken the opportunity to thank you for this amazing utility so far, so thank you!

Thank you, very appreciated! I'm also learning so much from you right now, that's a great opportunity for me as well!

**Comment by steven-omaha** [comment]
I think I found the issue. So far comparison of `Package`s was based on `PartialEq`. You introduced `ManagedBackend` which uses a `BTreeSet<Package>`, for which `BTreeSet<Package>::difference(...) where Package: Ord` is defined. So I think this uses `Ord`, which so far I have `derive`d in `src/grouping/package.rs:9`. If I understood this correctly you will need to `impl Ord for Package` to properly fix this bug.

**Comment by ripytide** [comment]
Great catch! I've made the `Ord`/`PartialEq`... all consistent with each other and refactored some other stuff.

**Comment by steven-omaha** [comment]
Thank you, merging

**Pull request merged**